lib/session.ts

```diff
@@ -0,0 +1,348 @@
+    }
+  }
+
+  /**
+   * Closes the amqp session.
+   * @return {Promise<void>} Promise<void>
+   * - **Resolves** the promise when rhea emits the "session_close" event.
+   * - **Rejects** the promise with an AmqpError when rhea emits the "session_error" event while trying
+   * to close an amqp session.
+   */
+  close(): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      log.error("[%s] The session is open ? -> %s", this.connection.id, this.isOpen());
+      if (this.isOpen()) {
+        let onError: Func<rhea.EventContext, void>;
+        let onClose: Func<rhea.EventContext, void>;
+        let waitTimer: any;
+
+        const removeListeners = () => {
+          clearTimeout(waitTimer);
+          this._session.removeListener(SessionEvents.sessionError, onError);
+          this._session.removeListener(SessionEvents.sessionClose, onClose);
+        };
+
+        onClose = (context: rhea.EventContext) => {
+          removeListeners();
+          process.nextTick(() => {
+            log.session("[%s] Resolving the promise as the amqp session has been closed.",
+              this.connection.id);
+            resolve();
+          });
+        };
+
+        onError = (context: rhea.EventContext) => {
+          removeListeners();
+          log.error("[%s] Error occurred while closing amqp session.",
+            this.connection.id, context.session!.error);
+          reject(context.session!.error);
+        };
+
+        const actionAfterTimeout = () => {
+          removeListeners();
+          const msg: string = `Unable to close the amqp session due to operation timeout.`;
+          log.error("[%s] %s", this.connection.id, msg);
+          reject(new Error(msg));
+        };
+
+        this._session.once(SessionEvents.sessionClose, onClose);
+        this._session.once(SessionEvents.sessionError, onError);
+        log.session("[%s] Calling session.close()", this.connection.id);
+        waitTimer = setTimeout(actionAfterTimeout, defaultOperationTimeoutInSeconds * 1000);
+        this._session.close();
+      } else {
+        resolve();
+      }
+    });
+  }
+
+  /**
+   * Creates an amqp receiver on this session.
+   * @param {Session} session The amqp session object on which the receiver link needs to be established.
+   * @param {ReceiverOptions} [options] Options that can be provided while creating an amqp receiver.
+   * @return {Promise<Receiver>} Promise<Receiver>
+   * - **Resolves** the promise with the Receiver object when rhea emits the "receiver_open" event.
+   * - **Rejects** the promise with an AmqpError when rhea emits the "receiver_close" event while trying
+   * to create an amqp receiver.
+   */
+  createReceiver(options?: ReceiverOptions): Promise<Receiver> {
+    if (options &&
+      ((options.onMessage && !options.onError) || (options.onError && !options.onMessage))) {
+      throw new Error("Both onMessage and onError handlers must be provided if one of " +
```

Comment:
this will throw synchronously rather than returning a rejected promise. if the caller of this function isn't using async/await, it will likely cause problems for them. you can fix it by:

* making `createReceiver` async
* moving this logic inside of the `new Promise()` below
* changing this to `return Promise.reject(new Error(...))`